File: 牛客网面试真题/字节跳动/2018第二批/3-二分查错/binary-search.js
/**
 * 正确写法：类似bisectRight
 */
function binarySearch (data, target) {
  let [left, right] = [0, data.length]

  while (left < right) {
    const mid = left + (right - left >>> 1)
    if (target < data[mid]) {
      right = mid
    } else {
      left = mid + 1
    }
  }

  // 此时left、right相等，随便用一个
  let res = -1
  if (data[right - 1] === target) {
    res = right - 1
  }
  return res
}

/**
 * 综上，题目所含错误有：
 * 1. 当`target < data[mid]时`，应该是`right = mid`，右边界为开区间
 * 2. 最终判断相等时，条件应该是`data[right-1] === target`，并且返回值也应该是`right-1`
 */

File: biweekly-contest/biweekly-contest-16/2-sum-of-mutated-array-closest-to-target/binary-search.cpp
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

class Solution {
public:
    // 使最大值最小，bisectLeft
    int findBestValue(vector<int>& arr, int target) {
        int sum = 0;
        int maximum = -1;
        for (int n : arr) {
            sum += n;
            maximum = max(maximum, n);
        }
        int diff = abs(sum - target);
        if (diff == 0) return maximum;

        // 二分
        int left = 1, right = maximum + 1;
        int mid;
        int currSum, prevSum;

        while (left < right) {
            mid = (left + right) / 2;

            // 检查mid
            currSum = getSum(arr, mid);
            prevSum = getSum(arr, mid - 1);

            if (prevSum >= target) {
                right = mid;
            } else if (currSum <= target) {
                left = mid + 1;
            } else {
                break;
            }
        }

        int curr = abs(currSum - target);
        int prev = abs(prevSum - target);
        if (prev <= curr) {
            return mid - 1;
        } else {
            return mid;
        }
    }

    int getSum(vector<int> &arr, int mid) {
        int res = 0;
        for (int n : arr) {
            res += min(n, mid);
        }
        return res;
    }
};